geaflow/geaflow-dsl/geaflow-dsl-connector/geaflow-dsl-connector-jdbc/src/main/java/com/antgroup/geaflow/dsl/connector/jdbc/JDBCTableSource.java

```diff
@@ -0,0 +1,162 @@
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JDBCTableSource implements TableSource {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JDBCTableSource.class);
+    private Configuration tableConf;
+    private StructType schema;
+    private String driver;
+    private String url;
+    private String username;
+    private String password;
+    private String tableName;
+    private Connection connection;
+    private Statement statement;
+
+    @Override
+    public void init(Configuration tableConf, TableSchema tableSchema) {
+        LOGGER.info("prepare with config: {}, \n schema: {}", tableConf, tableSchema);
+        this.tableConf = tableConf;
+        this.schema = tableSchema;
+
+        this.driver = tableConf.getString(JDBCConfigKeys.GEAFLOW_DSL_JDBC_DRIVER);
+        this.url = tableConf.getString(JDBCConfigKeys.GEAFLOW_DSL_JDBC_URL);
+        this.username = tableConf.getString(JDBCConfigKeys.GEAFLOW_DSL_JDBC_USERNAME);
+        this.password = tableConf.getString(JDBCConfigKeys.GEAFLOW_DSL_JDBC_PASSWORD);
+        this.tableName = tableConf.getString(JDBCConfigKeys.GEAFLOW_DSL_JDBC_TABLE_NAME);
+    }
+
+    @Override
+    public void open(RuntimeContext context) {
+        try {
+            Class.forName(this.driver);
+            this.connection = DriverManager.getConnection(url, username, password);
+            this.statement = connection.createStatement();
+        } catch (Exception e) {
+            throw new GeaFlowDSLException("failed to connect to database", e);
+        }
+    }
+
+    @Override
+    public List<Partition> listPartitions() {
+        return Collections.singletonList(new JDBCPartition(this.tableName));
+    }
+
+    @Override
+    public <IN> TableDeserializer<IN> getDeserializer(Configuration conf) {
+        return (TableDeserializer<IN>) new RowTableDeserializer();
+    }
+
+    @Override
+    public <T> FetchData<T> fetch(Partition partition, Optional<Offset> startOffset,
+                                  long windowSize) throws IOException {
+        if (!partition.getName().equals(this.tableName)) {
+            throw new GeaFlowDSLException("wrong partition");
+        }
+        long offset = 0;
+        if (startOffset.isPresent()) {
+            offset = startOffset.get().getOffset();
+        }
+        List<Row> dataList;
+        try {
+            dataList = JDBCUtils.selectRowsFromTable(this.statement, this.tableName,
+                this.schema.size(), offset, windowSize);
+        } catch (SQLException e) {
+            throw new GeaFlowDSLException("select rows form table failed.", e);
+        }
+        JDBCOffset nextOffset = new JDBCOffset(offset + windowSize);
```

Comment:
It is not necessarily guaranteed to read as much data as windowSize. If the table has been fully read, you need to set isFinish to true